app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,168 @@
+    # can be like this instead of in `after_perform`, the benefit is that we don't use undocumented order of callbacks
+    # return unless hierarchy.present?
+    #
+    # # depending on
+    # unlock(resource: lock_key)
+    # self.class.perform_later(*hierarchy)
   end
 
-  def destroyable_association?(_association)
-    true
+  # we can just use job.arguments.first but for compatibility mode we want it not to lock
+  def lock_key
+    first_argument = arguments.first
+    first_argument.is_a?(String) ? first_argument : Random.uuid
   end
 
   private
 
-  def called_from_provider_hierarchy?
-    return unless (tenant_id = object.tenant_id)
-    caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
+  def hierarchy_entries_for(...)
+    self.class.send(:hierarchy_entries_for, ...)
   end
 
-  def callback_options
-    { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
+  # handles a single hierarchy entry
+  # @return [Array<String>] hierarchy for a newly discovered object from association to delete or empty array otherwise
+  def handle_hierarchy_entry(entry)
+    case entry
+    when /Plain-([:\w]+)-(\d+)/
+      ar_object = $1.constantize.find($2.to_i)
+      # callbacks logic differs between object destroyed by association, it is standalone if first job arg is itself
+      ar_object.destroyed_by_association = DummyDestroyedByAssociationReflection.new(entry) if arguments.first != entry
+      ar_object.background_deletion_method_call
+      []
+    when /Association-([:\w]+)-(\d+):(\w+)/
+      handle_association($1.constantize.find($2.to_i), $3, entry)
+    else
+      raise ArgumentError, "Invalid entry specification: #{entry}"
+    end
+  rescue ActiveRecord::RecordNotFound => exception
+    Rails.logger.warn "#{self.class} skipping object, maybe something else already deleted it: #{exception.message}"
+  rescue NameError
+    raise DoNotRetryError, "seems like unexpectedly broken delete hierarchy entry: #{entry}"
   end
 
-  class ReflectionDestroyer
-
-    def initialize(main_object, reflection, caller_worker_hierarchy)
-      @main_object = main_object
-      @reflection = reflection
-      @caller_worker_hierarchy = caller_worker_hierarchy
+  # @return a single associated object for deletion or nil if non in the association
+  def handle_association(ar_object, association, hierarchy_association_string)
+    reflection = ar_object.class.reflect_on_association(association)
+    case reflection.macro
+    when :has_many
+      # here we keep original hierarchy entry if we still find an associated object
+      dependent = ar_object.public_send(association).public_send(:background_deletion_scope).take
+      dependent ? [hierarchy_association_string, *hierarchy_entries_for(dependent)] : []
+    when :has_one
+      # maximum of one associated so we never keep the original hierarchy entry
+      hierarchy_entries_for ar_object.public_send(association)
```

Comment:
```suggestion
      dependent = ar_object.public_send(association)
      dependent ? hierarchy_entries_for(dependent) : []
```

There is an issue with `:profile` association on Account, it is initialized lazily, so for some providers will not exist.